### PR TITLE
Dumb implementation for support --port and --host for hooks server

### PR DESCRIPTION
**Pull request by ddelnano**
#### :rocket: Why this change?

To allow compliant hooks servers to listen on different ports and ip addresses

#### :memo: Related issues and Pull Requests

https://github.com/apiaryio/dredd-hooks-template/issues/11
https://github.com/apiaryio/dredd-hooks-template/pull/17

#### :white_check_mark: What didn't I forget?

Still very much WIP since I believe there will need to be a discussion on what flags the hooks servers should look for (php already uses `--port` and `--host` but this can be up for discussion), how to stage these changes amongst the projects, etc.  I just wanted to get the simplest thing done so that I could run the dredd-hooks-php cucumber tests with [this](https://github.com/apiaryio/dredd-hooks-template/pull/17) new test to validate it works properly.

### CODE REVIEW

File: src/hooks-worker-client.coffee [hunk 1]
```diff
@@ -163,9 +163,10 @@ class HooksWorkerClient
         callback()
 
   spawnHandler: (callback) ->
-    pathGlobs = [].concat @runner.hooks?.configuration?.options?.hookfiles
+    pathGlobs = [].concat ["--port", "#{@handlerPort}", "--host", "#{@handlerHost}"], @runner.hooks?.configuration?.options?.hookfiles
```
Comment (honzajavorek) [inline src/hooks-worker-client.coffee]: Also, lately I tend to prefer to have all Dredd options in the `--option=value` format, consistently. In the case of Dredd itself, both formats should work, I just wanted to choose one way to go and stick with it.

File: src/hooks-worker-client.coffee [hunk 2]
```diff
@@ -163,9 +163,10 @@ class HooksWorkerClient
...
+    console.info pathGlobs
```
Comment (ddelnano) [inline src/hooks-worker-client.coffee]: Yep, removed it.

File: src/hooks-worker-client.coffee (hunk 1 above)
Comment (ddelnano) [inline src/hooks-worker-client.coffee]: Updated both of these.

**Comment by ddelnano** [comment]
## Todo
- [ ] Find out impact for existing hook servers
- [ ] Finish discussion on if this is what we want.  I like the idea of Dredd being able to discover if hooks support these flags at runtime but this is a much bigger change.

**Comment by ungrim97** [comment]
Will require an update to support the port/host arguments. Passing them in to the current version shouldn't break anything. (it will look for hookfiles files called --port=1234 and --host=127.0.0.1 but obviously won't find one and will skip them)

Will make an update to accept the new params

**Comment by ungrim97** [comment]
https://github.com/ungrim97/Dredd-Hooks/pull/7 should cover whats needed to make the Perl version work for this

**Comment by w-vi** [comment]
I'll check Python hooks and see if I can implement this feature quickly.

**Comment by ddelnano** [comment]
@honzajavorek @w-vi @ungrim97 so looks like all existing hooks servers will not be affected by this change except the Go ones.  How do we want to handle that?  I can get those updated soon but wanted to understand better what the plan was.

**Comment by w-vi** [comment]
Python hooks won't do anything bad, will skip the args so no issue here.

**Comment by ddelnano** [comment]
But what I mean is do we want to continue down this path?  It's going to cause existing go hook servers problems.  All other hooks seem fine but still.

**Comment by honzajavorek** [comment]
I'm sorry for not being prompt in answering (got sick). @ddelnano thanks very much for checking all the hooks! This is amazing work.

I see three options how to proceed:

- **Let the older Go hooks installations crash.** I don't like it very much, but I can't decide, since I have very little insight into how many people use the Go hooks, how prompt they are in upgrading to newer versions, etc. Probably up to @ddelnano and @snikch to consider.
- **The `features` event mentioned by @w-vi in https://github.com/apiaryio/dredd-hooks-template/pull/17#issuecomment-279988469** I like the idea, but implementing it would involve a lot of coding - both on hooks and Dredd sides. I'd prefer to have this as a future improvement instead of an instant solution.
- **Hardcoded check.** Is there a way we could hard code a simple check for Go hooks version? It's the least visually appealing solution, but doesn't involve much coding, doesn't make anything crash, and gets the job done.

What do you think?

**Comment by ddelnano** [comment]
@honzajavorek I definitely want to avoid option 1 :).  And since the `features` event seems like alot more work than what we need for now some type of hard coded check will be necessary.  I am going to brainstorm some ideas the rest of today and post back some potential ideas.

**Comment by ddelnano** [comment]
@honzajavorek sorry for the long delay, had an unexpected hospital trip but I'm doing much better.

Another way we could check the hooks server that I thought of that would be very dumb but get the job done.  We could have dredd try and run the hooks server with the `--host` and `--port` arguments and if it fails try it without the `--host` and `--port` flags.  The drawback is that unrelated failures will probably fail twice.  For the hardcoded version check on the Go hooks how would we accomplish that?  There is no way to check the version using the go hook binary.

**Comment by honzajavorek** [comment]
@ddelnano Whoa! 🚑 Whatever it was, I wish you to get better quickly!

The fact there is no way we can hardcode a check for go hooks version is unfortunate. I'll think about the trial & error approach. At first sight, I don't like it very much, but at the same time, it can be the only way we're left with to save the situation 🙂

**Comment by ddelnano** [comment]
@honzajavorek thanks, things are back to normal now :).

One thing I didn't think of for the hardcoded check.  In the next version of the go hooks we could add a command to output the version and then if it doesn't support the version command we can assume that it won't support the `--host` and `--port`.

**Comment by honzajavorek** [comment]
@ddelnano What is the current behavior of go hooks if they're provided with `--version`? Do they start and wait for connection, do they crash, ...?

**Comment by ddelnano** [comment]
It's going to crash...

**Comment by honzajavorek** [comment]
Then I think we're okay with what you proposed in https://github.com/apiaryio/dredd/pull/712#issuecomment-287864558. Dredd could try to run the hook handler with options and if it crashes, it tries to run it without the options. If it doesn't crash, it respects the test results.

**Comment by ddelnano** [comment]
Ok. Yea obviously not ideal but there aren't many options unfortunately. This weekend I will have some time I'll give that a shot. And work on getting the go hooks updated to use the new flags.

**Comment by honzajavorek** [comment]
I'm very sorry we were not able to get back to this. I need to close this as we're going to work on #705 first, which requires having no pending Pull Requests. We'll get back to this proposal as soon as we're addressing #917. Finishing this PR is listed on #917 as one of the necessary steps, so it shouldn't get forgotten. Thank you for the contribution! I hope we'll have the chance to tidy up the template soon.

**Pull request closed without merging**